Per-package transpilation

BinaryMuse opened the pull request:
This is a ~~WIP and (thus far) rough~~ implementation of per-package transpilation pipelines mentioned in #13087. It allows packages to opt in to fully managing their own transpilation steps, bypassing the built-in Atom transpilation entirely. Here's roughly how it works:
1. A package declares a property in its `package.json` called `atomTranspilers`. This is an array of objects, with each object specifying `glob`, `transpiler`, and optionally `options`.
2. When a package is loaded, if `atomTranspilers` exists in its metadata, we notify the `CompileCache` that it wants to do its own compilation.
3. A new class called `PackageTranspilationRegistry` wraps all the `CompileCache` compilers with a new object of the same interface that falls back to the original compiler if a package hasn't opted in to its own transpilation, but bypasses it completely otherwise.
4. If a package is doing it's own compilation, then for each file that matches the `glob` specified, the associated `transpiler` is required, and the exported value should be an object with a function called `compile` that takes three arguments: `sourceCode`, `filePath`, and `options` (where `options` comes from the package's `package.json`). It should return a string or Buffer with the transpiled source, or `undefined` to indicate that it doesn't want to transpile that file. Note that returning `undefined` doesn't fall back to the legacy, core behavior — packages which opt in to this mechanism are _fully_ responsible for _all_ their transpilation.

Also worth noting is that the transpiler file cannot be transpiled itself; its source must run in a vanilla Atom environment (which is admittedly pretty capable these days :).

One question I have is how to handle the caching. For example, I currently have `getCachePath` set to

``` javascript
return path.join(
  "package-transpile",
  crypto
    .createHash('sha1')
    .update(transpilerSource, 'utf8')
    .update(sourceCode, 'utf8')
    .digest('hex')
  )
```

but in many cases it's possible that neither the file's source nor the transpiler's source has changed, but the output of transpilation would be different (e.g. when editing a `.babelrc` file). Perhaps we should allow the package author to also specify a cache key for the file?

/cc @nathansobo @atom/core for early 💭

Commit:
First pass at per-package transpilation

Commit:
Pass config options to package transpilers

Commit:
Implement glob checking and multiple transpilers

Commit:
:fire: transpilerHashes

Commit:
:fire: console

Commit:
Remove lookup in favor of passed arg

Commit:
Fix linting issues/errors

nathansobo commented:
Excited to see this coming together! I wonder if we can expose something that can be an ingredient to the cache key like a version number that we can hash together with the options. I guess giving them direct control is okay but it would be nice to be able to make our own solid guarantees about the performance of the cached code path.

nathansobo commented:
A few more thoughts that bubbled into my head...

* Should the transpiler module's method be called `transpile` rather than `compile` for consistency? Or should we go the other way and name the key `compilers`?
* Refining my thoughts on the participation in cache invalidation... I'm not sure a version is enough because transpilers may want to support mechanisms such as `.babelrc`, the contents of which can change. So perhaps it's safest to just offer maximal power with an optional `getCacheKey` method and hope people don't abuse it.

Finally some broader thoughts that may extend beyond the scope of this PR but I wanted to bring them up? Do we need to prefix our manifest fields with `atom`? I'm actually not clear on where else we do this, but considering this `package.json` is actually for an Atom package and not a node module, it seems like it's our namespace to populate. If this were intended to be a vanilla node module I'd feel differently, but since these things are going to run in Atom... If we do decide to prefix, I'd be happier with a single `atom` sub-object containing all the Atom-related stuff.

Commit:
:fire: Extra // in eslint comment

BinaryMuse commented:
@nathansobo 

> Should the transpiler module's method be called `transpile` rather than `compile` for consistency? Or should we go the other way and name the key `compilers`?

Good point. "Transpile" is such a weird word, I'd almost just rather use "compile" in user-facing stuff.

> So perhaps it's safest to just offer maximal power with an optional `getCacheKey` method and hope people don't abuse it.

This feels reasonable. People can already do plenty of stuff to fuck up performance, and ideally if anyone abuses it it'll be picked up by timecop and fixed in a PR. 😁 

> Do we need to prefix our manifest fields with `atom`?

I did this partly out of general safety (yes, it's an Atom package, but other Node things that people use stuff shit in `package.json`) and consistency (we already have `atomTestRunner`). I kind of like the idea of a single `atom` sub-object, but it feels like it's too late for that?

Commit:
Include user data in cache calculation

Commit:
compile :arrow_right: transpile

BinaryMuse commented:
Okay, I went with "transpile" because it's the language we use everywhere else. I'm not attached, though.

For the caching, I added an optional method to the transpiler called `getCacheKeyData` that allows the user to return *additional* data that helps determine how a file is cached; for example, in a recent package, I used the contents of all the files that affect transpilation:

```javascript
var cacheKeyData = [
  fs.readFileSync(path.join(__dirname, '.babelrc')),
  fs.readFileSync(path.join(__dirname, 'schema.json')),
  fs.readFileSync(path.join(__dirname, 'relay-babel.js'))
].join('\n')

module.exports = {
  getCacheKeyData: function () {
    return cacheKeyData
  },

  // ...
```

Commit:
Call  on transpiled code

Commit:
Throw when can't find custom transpiler

Commit:
Add PackageTranspilationRegistry spec

BinaryMuse commented:
A thought I had as I was working on specs for this: this will only work for `.js`, `.ts`, and `.coffee` files because those are the only three with `require.extensions` hooks. We should probably allow users to specify other extensions they want to load; this would install a "pass-through" compiler on those extensions for files that aren't transpiled by the pipeline. One thought was to add an explicit `extensions` option in the `package.json` bit; it's a little repetitive since many `glob`s will already have an extension listed, but I can't think of a good way to handle it in a generic case.

Commit:
Move call to fs.realpathSync to CompileCache

BinaryMuse commented:
So, I was wrong — obscure extensions work fine because Node defaults all unknown extensions to the `require.extensions['.js']` compiler, so as long as we keep that one assigned to a `PackageTranspilationRegistry` wrapped compiler, we're good to go.

---

This works perfectly with an internal package I'm working on and doesn't seem to add any significant startup time to the package. I think it's ready for more serious 👀 now. In particular, I'd love feedback on places where lookups and the like are less than optimal, or places where folks experience slowdowns when loading Atom.

Failing tests on appveyor are unrelated.

/cc @atom/feedback

BinaryMuse commented:
Wondering if there should be some code to specifically ignore `node_modules`; `**/*.js` is likely to be a common glob. Are there any instances where the user would *want* to ensure `node_modules` is included?

nathansobo commented:
I think we want to always exclude `node_modules`.

nathansobo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,151 @@
+      if (config = this.configByPackagePath[thisPath]) { // eslint-disable-line no-cond-assign
+        for (var i = 0; i < config.specs.length; i++) {
+          spec = config.specs[i]
+          if (minimatch(filePath, path.join(config.path, spec.glob))) {
```
Maybe you want to compile the minimatch regexp up front? Not sure if they do caching internally.

BinaryMuse replied:
Hm, doesn't look like minimatch actually uses a regex under the hood.

nathansobo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,151 @@
+      hash.update(additionalCacheData, 'utf8')
+    }
+
+    return path.join('package-transpile', hash.digest('hex'))
```
I wonder if it's worth grouping transpiled sources in their own subfolder named after the package... that might make it easier to delete them when debugging or something. I don't have a super strong opinion on this one but I thought I'd raise the idea.

BinaryMuse replied:
This was a good idea, thanks. 👍

nathansobo commented:
It seems like we only support the extensions `.js`, `.ts` and `.coffee`... am I missing something? If this is true, think you should add probably have some facility for declaring a new extension so this is fully general.

BinaryMuse commented:
> I think we want to always exclude `node_modules`.

👍 

> Maybe you want to compile the minimatch regexp up front?

Good thought, I'll look into that.

> I wonder if it's worth grouping transpiled sources in their own subfolder named after the package

I considered this but wasn't sure. I feel more confident now that you've mentioned it. I'll try it and see how it feels.

> It seems like we only support the extensions `.js`, `.ts` and `.coffee`... am I missing something?

Yeah, Node [falls back to `.js`](https://github.com/nodejs/node/blob/40366df885ec75c7eeee5e7e7626212ae1a6e770/lib/module.js#L489-L490) if it doesn't recognize the extension.

Commit:
Expect transpilation result on .code

Commit:
We don't want no node_modules

Commit:
Add test for non-standard extension transpilation

BinaryMuse commented:
Okay, ignoring `node_modules` is in, and I modified the transpilation function to expect the transpiled source code on a `.code` property so we have room to do fancy stuff with source maps and stuff in the future (for now source maps will work if you just add an inline one to the transpiled source code). I also added a test case for file extensions that aren't registered with `require.extensions`. The only problem would be if some package takes over some extension (like `.jsx`) and then another package tries to transpile on that extension.

Commit:
Finish up a few missing expectations

Commit:
:shirt:

Commit:
Include package name as part of cache key

BinaryMuse commented:
I believe this is ready to go.

thomasjo commented on spec/package-transpilation-registry-spec.js:
```diff
@@ -0,0 +1,145 @@
+
+import PackageTranspilationRegistry from '../src/package-transpilation-registry'
+
+let originalCompiler = {
```
`const`?

thomasjo commented on spec/package-transpilation-registry-spec.js:
```diff
@@ -0,0 +1,145 @@
+      getCacheKeyData: (sourceCode, filePath, options) => {
+        return 'omg-transpiler-cache-data'
+      }
+    }
```
Looks like all declarations above can be `const`-ified.

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+var path = require('path')
+
+var minimatch = require('minimatch')
+
```
`const`-ify all the above

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+
+var minimatch = require('minimatch')
+
+var Resolve = null
```
`let`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+  // that falls back to the original transpiler implementation if and
+  // only if a package hasn't registered its desire to transpile its own source.
+  wrapTranspiler: function (transpiler) {
+    var self = this
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+    var self = this
+    return {
+      getCachePath: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+      },
+
+      compile: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+        if (self.transpilerPaths[filePath]) {
+          return false
+        }
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+    var config = null
+    var spec = null
+    var thisPath = filePath
+    var lastPath = null
```
`let`-ify all these

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+    // only iterates four times, even if there are hundreds of configs registered.
+    while (thisPath !== lastPath) { // until we reach the root
+      if (config = this.configByPackagePath[thisPath]) { // eslint-disable-line no-cond-assign
+        for (var i = 0; i < config.specs.length; i++) {
```
`let`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+      .update(JSON.stringify(spec.options || {}))
+      .update(transpilerSource, 'utf8')
+      .update(sourceCode, 'utf8')
+
```
Use `const` and `let` as appropriate above.

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+      .update(transpilerSource, 'utf8')
+      .update(sourceCode, 'utf8')
+
+    var additionalCacheData
```
`let`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+    var transpiler = this.getTranspiler(spec)
+
+    if (transpiler) {
+      var result = transpiler.transpile(sourceCode, filePath, spec.options || {})
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+  },
+
+  transpileWithPackageTranspiler: function (sourceCode, filePath, spec) {
+    var transpiler = this.getTranspiler(spec)
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+        throw new Error('Could not find a property `.code` on the transpilation results of ' + filePath)
+      }
+    } else {
+      var err = new Error("Could not resolve transpiler '" + spec.transpiler + "' from '" + spec._config.path + "'")
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+  },
+
+  getTranspiler: function (spec) {
+    var transpilerPath = this.getTranspilerPath(spec)
```
`const`

thomasjo commented on src/package-transpilation-registry.js:
```diff
@@ -0,0 +1,159 @@
+  getTranspiler: function (spec) {
+    var transpilerPath = this.getTranspilerPath(spec)
+    if (transpilerPath) {
+      var transpiler = require(transpilerPath)
```
`const`

thomasjo reviewed (dismissed):
Nitpicking related to lack of `const` and `let` usage.

BinaryMuse commented:
@thomasjo Following the style here: https://github.com/atom/atom/blob/daf790406f42fee77355fed56ece280ac6e60c10/src/compile-cache.js#L3-L6

I'll change the ones in the spec file.

Commit:
const-ify some lets

thomasjo commented:
@BinaryMuse That code is almost a year old though. We have native support for most of ES6/7 now. I really think we should try to ensure new code is as modern as possible. I'll happily modernize `compile-cache.js` if necessary.

Commit:
Modernify PackageTranspilationRegistry

Commit:
:shirt:

nathansobo commented:
@BinaryMuse Shoot one thing I forgot. Could you document this in the Flight Manual or some other appropriate spot?

BinaryMuse commented:
@nathansobo Already on it :)

maxbrunsfeld commented:
👏

zertosh commented:
I totally spaced on this PR. The thing that worries me about this, is that it encourages package authors to ship (1) untranspiled code, (2) put babel & friends in `dependencies` (and bloating packages)

BinaryMuse commented:
@zertosh After some discussion, we decided to go with this solution. We don't have a facility to allow package authors to ship transpiled code easily; they'd have to check it into their repository, making diffs and pull requests more difficult. Additionally, we still wanted to have a mechanism for at-runtime transpilaton when in dev mode. The package bloat is a concern, but given the circumstances (and the fact that we want to do global deduping of installed modules), we were willing to eat the cost.

zertosh commented:
> they'd have to check it into their repository, making diffs and pull requests more difficult.

Yes, but it actually works out if you follow a different publish flow from just running `apm publish` from `master`. We publish Nuclide pre-transpiled. We first transpile everything, commit the code (but we don't merge it into `master`), we tag it (`git tag vX.X.X`), and then publish the tag `apm publish --tag vX.X.X`). So master never has transpiled code - so no conflicts or anything.

> The package bloat is a concern

Bloat is one. Another one is load time. If package authors are not super careful about lazy loading their transpiler, you're going to end up paying a big perf penalty. Babel is not a small dependency, it can take upwards of 50ms to just `require` it.

nathansobo commented:
I think this should probably only work in development mode and we should fix the inability to publish compiled package data.

BinaryMuse commented:
> If package authors are not super careful about lazy loading their transpiler, you're going to end up paying a big perf penalty.

Package authors will not be required to require their transpiler as part of their package code path. It's a separate module that the Atom internals require if and only if the given files are not already cached (so on first load of a package and after any package update).

> what if we only enabled this in dev mode and fix the root problem?

I like the idea in theory, though in practice it would require different code/steps for running in dev and compiling for publishing. That said, I'd be open to it if we can make it work smoothly.

nathansobo commented:
Yeah, I suppose you're right about the caching meaning we'll avoid requiring the transpiler in most cases. After digesting that point again, I'm open to the chance of a one-time slowdown to get stuff transpiled on install. Ultimately, I do think it makes the most sense to publish transpiled assets, but that's obviously uncontroversial it's just a lot of work.

zertosh commented:
> Package authors will not be required to require their transpiler as part of their package code path. It's a separate module that the Atom internals require if and only if the given files are not already cached 

Hmm, maybe I'm misreading the code, but it looks like you still require `atomTranspilers.transpiler` to check for a `getCacheKeyData` function. So if you put your requires at the top of the `atomTranspilers.transpiler` file (as most people do), you're going to pay the load cost for `babel` or whatever every time - unless the package author inline's the `require` call.

Package authors would have to write something like this to avoid the penalty:

```js
let version;
let babel;

module.exports = {
  transpile(sourceCode, filePath, options) {
    if (babel == null) babel = require('babel-core');
    return babel.transpile(/*....*/);
  },

  getCacheKeyData(sourceCode, filePath, options) {
    // or whatever you want to base the cache key on.
    if (version == null) version = require('babel-core/package.json').version;
    return version;
  }  
}
```

Side note: We use https://github.com/zertosh/babel-plugin-transform-inline-imports-commonjs in Nuclide and in yarn to inline imports to defer `require` calls until the imported identifier is referenced.

nathansobo commented:
I'm imagining package authors relying on libraries to do the transpiler support. So something like `atom-babel` would implement the transpiler object and you'd just reference that in your `node_modules`. @BinaryMuse that makes me realize i didn't check whether the transpiler path is passed to `require.resolve` like we do with the test runner.

zertosh commented:
Is `atom-babel` a hypothetical package? I couldn't find it lol.. Maybe it's my trauma talking but the majority of transpile setups I've seen are pretty bad. In any case, I left my 2c.

nathansobo commented:
Yeah, your 2c is super valuable and we appreciate it. I think at this point we're going to risk moving forward because we need this on a package we're working on and don't have time to setup the publishing flow. Long term you're absolutely right that we should just transpile on publish.

BinaryMuse commented:
> Hmm, maybe I'm misreading the code, but it looks like you still require `atomTranspilers.transpiler` to check for a `getCacheKeyData` function.

Ah, yes, you're right of course. I forgot about that detail. I like giving user full control over generating (a portion of) the cache key, but I wonder if it'd be better to just allow the user to specify a list of files that go into generation of the key so we can just pull stuff from `package.json` without needing to require the module?

Or, as @nathansobo said, point people toward pre-made modules (I'm working on one for Babel 6 now) that implement transpilation for the user, and highlight the edge cases in the docs for people that want to make their own.

BinaryMuse commented:
> that makes me realize i didn't check whether the transpiler path is passed to `require.resolve` like we do with the test runner.

Yes, `getTranspilerPath` [runs the path through `Resolve.sync`](https://github.com/atom/atom/pull/13101/files#diff-42678e5bce81ab920184a4a8415043bcR143).

Arcanemagus commented:
Has this been added to the Flight Manual yet? A search for `atomTranspilers` shows no results at the moment.

BinaryMuse commented:
@Arcanemagus It hasn't; I've been ironing out the details and wrinkles as I've used the feature, and will be adding it once it's ready.

danielbayley commented:
> only support the extensions .js, .ts and .coffee

So am I right in thinking…
1. This can't be utilised to take advantage of PostCSS in a say theme? (See also #5076 for a more general idea on that…)
2. This is effectively a temporary patch to address the root problem which is basically that `apm` doesn't work like `npm` when it comes to publishing…

BinaryMuse commented:
@danielbayley 

> This can't be utilised to take advantage of PostCSS in say, a theme?

This is all about the `CompileCache` in Atom, which is the mechanism which transpiles files via the Node.js `require` mechanism. So it would not be useful for CSS unless you ran it through Node's `require` somehow.

> This is effectively a temporary patch to address the root problem

It's not the solution we think is most ideal, but there certainly isn't any guarantee of or timeline for an alternative solution.
